src/area-textlayout.c

```diff
@@ -0,0 +1,58 @@
+#include <ui.h>
+#include "napi_utils.h"
+#include "control.h"
+#include "events.h"
+#include "values.h"
+
+static const char *MODULE = "DrawTextLayout";
+
+// font-descriptor.c
+typedef struct {
+	uiFontDescriptor *font;
+	bool fromButton;
+} FontHandle;
+
+static void free_layout(napi_env env, void *finalize_data, void *finalize_hint) {
+	uiDrawTextLayout *layout = (uiDrawTextLayout *)finalize_data;
+	uiDrawFreeTextLayout(layout);
+}
+
+LIBUI_FUNCTION(create) {
+	INIT_ARGS(4);
+	ARG_POINTER(uiAttributedString, str, 0);
+	ARG_POINTER(FontHandle, h, 1);
+	ARG_DOUBLE(width, 2);
+	ARG_INT32(align, 3);
+
+	// ok with memory?
```

Comment:
What happens if the string or font gets freed after initializing the layout? Should we keep a reference to prevent gc (napi_ref ?) ?

Reply:
Yes, it should be ideal to keep a reference, we can after release it e.g. in free_layout

Reply:
Like that?

```c
// create
napi_create_reference(env, str_value, 1, layout_handle->str_ref);
napi_create_reference(env, font_value, 1, layout_handle->font_ref);

// free
napi_delete_reference(env, layout_handle->str_ref);
napi_delete_reference(env, layout_handle->font_ref);

```

Reply:
I don't know if napi_delete_reference decrease the reference count or only deallocates it. I've always used https://nodejs.org/api/n-api.html#n_api_napi_reference_unref to decrease the ref count to 0.

> napi_create_reference(env, str_value, 1, layout_handle->str_ref);

I don't know if you can create a reference of a string, maybe you have to use https://nodejs.org/api/n-api.html#n_api_napi_coerce_to_object (I'm not sure)

Reply:
The string is a attributedstring:
`ARG_POINTER(uiAttributedString, str, 0);`